File: backend-node/src/routes/copyGeneration.ts
import { Router } from "express";

import { settings } from "../core/config";
import { asyncHandler } from "../middleware/asyncHandler";
import { validateBody } from "../middleware/validate";
import {
  CopyGenerationRequestInput,
  FieldRegenerationRequestInput,
  copyGenerationRequestSchema,
  fieldRegenerationRequestSchema,
} from "../schemas/copyGeneration";
import * as copyGenerationService from "../services/copyGenerationService";

const router = Router();

router.post(
  "/generate",
  validateBody(copyGenerationRequestSchema),
  asyncHandler(async (req, res) => {
    if (!settings.ANTHROPIC_API_KEY) {
      res.status(500).json({ detail: "Anthropic API key not configured" });
      return;
    }
    const body = req.body as CopyGenerationRequestInput;
    try {
      const result = await copyGenerationService.generateVariations({
        brand: body.brand,
        product: body.product,
        profile: body.profile,
        campaignDetails: body.campaignDetails,
        template: body.template,
        variationCount: body.variationCount,
        customPrompt: body.customPrompt,
      });
      res.json(result);
    } catch (err) {
      if (err instanceof SyntaxError) {
        res.status(500).json({ detail: `Failed to parse AI response as JSON: ${err.message}` });
        return;
      }
      res.status(500).json({ detail: `Copy generation failed: ${(err as Error).message}` });
    }
  })
);

router.post(
  "/regenerate-field",
  validateBody(fieldRegenerationRequestSchema),
  asyncHandler(async (req, res) => {
    if (!settings.ANTHROPIC_API_KEY) {
      res.status(500).json({ detail: "Anthropic API key not configured" });
      return;
    }
    const body = req.body as FieldRegenerationRequestInput;
    try {
      const newValue = await copyGenerationService.regenerateField(body);
      res.json({ newValue });
    } catch (err) {
      res.status(500).json({ detail: `Field regeneration failed: ${(err as Error).message}` });
    }
  })
);

export default router;

File: backend/app/api/v1/copy_generation.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
import json

from app.core.config import settings
from app.services import copy_generation_service

router = APIRouter()

class CopyGenerationRequest(BaseModel):
    brand: Dict[str, Any]
    product: Dict[str, Any]
    profile: Dict[str, Any]
    template: Optional[Dict[str, Any]] = None
    variationCount: int = 3
    campaignDetails: Dict[str, str]
    customPrompt: Optional[str] = None

class FieldRegenerationRequest(BaseModel):
    field: str
    currentValue: str
    brand: Dict[str, Any]
    product: Dict[str, Any]
    profile: Dict[str, Any]
    template: Optional[Dict[str, Any]] = None
    campaignDetails: Dict[str, str]

@router.post("/generate")
async def generate_copy(request: CopyGenerationRequest):
    """Generate direct-response ad copy variations using the doctrine-grounded copy engine"""

    if not settings.ANTHROPIC_API_KEY:
        raise HTTPException(status_code=500, detail="Anthropic API key not configured")

    try:
        result = copy_generation_service.generate_variations(
            brand=request.brand,
            product=request.product,
            profile=request.profile,
            campaign_details=request.campaignDetails,
            template=request.template,
            variation_count=request.variationCount,
            custom_prompt=request.customPrompt,
        )
        return result

    except json.JSONDecodeError as e:
        print(f"JSON Parse Error: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to parse AI response as JSON: {str(e)}")
    except Exception as e:
        print(f"Copy generation error: {e}")
        raise HTTPException(status_code=500, detail=f"Copy generation failed: {str(e)}")

@router.post("/regenerate-field")
async def regenerate_field(request: FieldRegenerationRequest):
    """Regenerate a specific field (headline, body, or cta)"""

    if not settings.ANTHROPIC_API_KEY:
        raise HTTPException(status_code=500, detail="Anthropic API key not configured")

    try:
        new_value = copy_generation_service.regenerate_field(
            field=request.field,
            current_value=request.currentValue,
            brand=request.brand,
            product=request.product,
            profile=request.profile,
            campaign_details=request.campaignDetails,
        )
        return {"newValue": new_value}

    except Exception as e:
        print(f"Field regeneration error: {e}")
        raise HTTPException(status_code=500, detail=f"Field regeneration failed: {str(e)}")
